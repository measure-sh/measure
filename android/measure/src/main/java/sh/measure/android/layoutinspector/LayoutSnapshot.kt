package sh.measure.android.layoutinspector

import sh.measure.android.MsrAttachment
import sh.measure.android.events.Attachment
import sh.measure.android.events.AttachmentType
import sh.measure.android.gestures.DetectedGesture
import sh.measure.android.tracing.InternalTrace

internal class LayoutSnapshot(private val nodes: List<Node>) {
    /**
     * Generates an SVG visualization of the layout snapshot as an attachment.
     *
     * The SVG includes all nodes in the layout with their positions and dimensions.
     * If a [targetNode] is provided, it will be highlighted in the visualization.
     *
     * @param targetNode Optional node to highlight in the SVG visualization
     * @param width The width of the SVG to be generated.
     * @param height The height of the SVG to be generated.
     * @return An [Attachment] containing the SVG data with MIME type set to LAYOUT_SNAPSHOT
     */
    fun generateSvgAttachment(targetNode: Node? = null, width: Int, height: Int): Attachment {
        return InternalTrace.trace(
            label = { "msr-generateSvgAttachment" },
            block = {
                val svg = nodes.generateSvg(targetNode, width, height)
                Attachment(
                    "snapshot.svg",
                    AttachmentType.LAYOUT_SNAPSHOT,
                    svg.encodeToByteArray(),
                )
            },
        )
    }

    /**
     * Generates an SVG visualization of the layout snapshot as an attachment.
     *
     * The SVG includes all nodes in the layout with their positions and dimensions.
     * If a [targetNode] is provided, it will be highlighted in the visualization.
     *
     * @param targetNode Optional node to highlight in the SVG visualization
     * @param width The width of the SVG to be generated.
     * @param height The height of the SVG to be generated.
     * @return An [Attachment] containing the SVG data with MIME type set to LAYOUT_SNAPSHOT
     */
    fun generateSvgMsrAttachment(targetNode: Node? = null, width: Int, height: Int): MsrAttachment {
        return InternalTrace.trace(
            label = { "msr-generateSvgAttachment" },
            block = {
                val svg = nodes.generateSvg(targetNode, width, height)
                MsrAttachment(
                    "snapshot.svg",
                    svg.encodeToByteArray(),
                    AttachmentType.LAYOUT_SNAPSHOT,
                )
            },
        )
    }

    fun findTargetNode(): Node? {
        return nodes.findLast { it.willConsumeGesture }
    }

    fun isEmpty(): Boolean {
        return nodes.isEmpty()
    }
}

internal data class Node(
    val id: String?,
    val className: String,
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
    val willConsumeGesture: Boolean = false,
    val gesture: DetectedGesture? = null,
    val isText: Boolean = false,
)
