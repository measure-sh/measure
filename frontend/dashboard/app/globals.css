@import 'tailwindcss';
@import "tw-animate-css";
/*
  ---break---
*/
@custom-variant dark (&:is(.dark *));

@config '../tailwind.config.ts';

/*
  The default border color has changed to `currentColor` in Tailwind CSS v4,
  so we've added these compatibility styles to make sure everything still
  looks the same as it did with Tailwind CSS v3.

  If we ever want to remove these styles, we need to add an explicit border
  color utility to any element that depends on these defaults.
*/
@layer base {

  *,
  ::after,
  ::before,
  ::backdrop,
  ::file-selector-button {
    border-color: var(--color-gray-200, currentColor);
  }
}

@layer utilities {
  :root {
    --foreground-rgb: 0, 0, 0;
    --background-start-rgb: 255, 255, 255;
    --background-end-rgb: 255, 255, 255;
  }

  button,
  [role='button'] {
    cursor: auto;
  }

  body {
    color: rgb(var(--foreground-rgb));
    background: linear-gradient(to bottom,
        transparent,
        rgb(var(--background-end-rgb))) rgb(var(--background-start-rgb));
  }
}

@layer base {
  html {
    scroll-behavior: smooth;
  }
}

/*
  ---break---
*/

@theme inline {
  --radius-sm: calc(var(--radius) - 4px);
  --radius-md: calc(var(--radius) - 2px);
  --radius-lg: var(--radius);
  --radius-xl: calc(var(--radius) + 4px);
  --color-background: var(--background);
  --color-foreground: var(--foreground);
  --color-card: var(--card);
  --color-card-foreground: var(--card-foreground);
  --color-popover: var(--popover);
  --color-popover-foreground: var(--popover-foreground);
  --color-primary: var(--primary);
  --color-primary-foreground: var(--primary-foreground);
  --color-secondary: var(--secondary);
  --color-secondary-foreground: var(--secondary-foreground);
  --color-muted: var(--muted);
  --color-muted-foreground: var(--muted-foreground);
  --color-accent: var(--accent);
  --color-accent-foreground: var(--accent-foreground);
  --color-destructive: var(--destructive);
  --color-border: var(--border);
  --color-input: var(--input);
  --color-ring: var(--ring);
  --color-chart-1: var(--chart-1);
  --color-chart-2: var(--chart-2);
  --color-chart-3: var(--chart-3);
  --color-chart-4: var(--chart-4);
  --color-chart-5: var(--chart-5);
  --color-sidebar: var(--sidebar);
  --color-sidebar-foreground: var(--sidebar-foreground);
  --color-sidebar-primary: var(--sidebar-primary);
  --color-sidebar-primary-foreground: var(--sidebar-primary-foreground);
  --color-sidebar-accent: var(--sidebar-accent);
  --color-sidebar-accent-foreground: var(--sidebar-accent-foreground);
  --color-sidebar-border: var(--sidebar-border);
  --color-sidebar-ring: var(--sidebar-ring);
}

/*
  ---break---
*/

:root {
  --radius: 0.625rem;
  --background: oklch(1 0 0);
  --foreground: oklch(1.145 0 0);
  --card: oklch(1 0 0);
  --card-foreground: oklch(0.145 0 0);
  --popover: oklch(1 0 0);
  --popover-foreground: oklch(0.145 0 0);
  --primary: oklch(0.205 0 0);
  --primary-foreground: oklch(0.985 0 0);
  --secondary: oklch(0.97 0 0);
  --secondary-foreground: oklch(0.205 0 0);
  --muted: oklch(0.97 0 0);
  --muted-foreground: oklch(0.556 0 0);
  --accent: oklch(94.5% 0.129 101.54);
  --accent-foreground: oklch(0.205 0 0);
  --destructive: oklch(0.577 0.245 27.325);
  --border: oklch(0.922 0 0);
  --input: oklch(0.922 0 0);
  --ring: oklch(90.5% 0.182 98.111);
  --chart-1: oklch(0.646 0.222 41.116);
  --chart-2: oklch(0.6 0.118 184.704);
  --chart-3: oklch(0.398 0.07 227.392);
  --chart-4: oklch(0.828 0.189 84.429);
  --chart-5: oklch(0.769 0.188 70.08);
  --sidebar: oklch(1 0 0);
  --sidebar-foreground: oklch(0.145 0 0);
  --sidebar-primary: oklch(0.205 0 0);
  --sidebar-primary-foreground: oklch(0.985 0 0);
  --sidebar-accent: oklch(94.5% 0.129 101.54);
  --sidebar-accent-foreground: oklch(0.205 0 0);
  --sidebar-border: oklch(0.922 0 0);
  --sidebar-ring: oklch(90.5% 0.182 98.111);
}

.dark {
  --radius: 0.625rem;
  --background: oklch(1 0 0);
  --foreground: oklch(1.145 0 0);
  --card: oklch(1 0 0);
  --card-foreground: oklch(0.145 0 0);
  --popover: oklch(1 0 0);
  --popover-foreground: oklch(0.145 0 0);
  --primary: oklch(0.205 0 0);
  --primary-foreground: oklch(0.985 0 0);
  --secondary: oklch(0.97 0 0);
  --secondary-foreground: oklch(0.205 0 0);
  --muted: oklch(0.97 0 0);
  --muted-foreground: oklch(0.556 0 0);
  --accent: oklch(94.5% 0.129 101.54);
  --accent-foreground: oklch(0.205 0 0);
  --destructive: oklch(0.577 0.245 27.325);
  --border: oklch(0.922 0 0);
  --input: oklch(0.922 0 0);
  --ring: oklch(90.5% 0.182 98.111);
  --chart-1: oklch(0.646 0.222 41.116);
  --chart-2: oklch(0.6 0.118 184.704);
  --chart-3: oklch(0.398 0.07 227.392);
  --chart-4: oklch(0.828 0.189 84.429);
  --chart-5: oklch(0.769 0.188 70.08);
  --sidebar: oklch(1 0 0);
  --sidebar-foreground: oklch(0.145 0 0);
  --sidebar-primary: oklch(0.205 0 0);
  --sidebar-primary-foreground: oklch(0.985 0 0);
  --sidebar-accent: oklch(94.5% 0.129 101.54);
  --sidebar-accent-foreground: oklch(0.205 0 0);
  --sidebar-border: oklch(0.922 0 0);
  --sidebar-ring: oklch(90.5% 0.182 98.111);
}



/*
  ---break---
*/

@layer base {
  * {
    @apply border-border outline-ring/50;
  }

  body {
    @apply bg-background text-foreground;
  }
}