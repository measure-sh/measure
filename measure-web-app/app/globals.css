@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --foreground-rgb: 0, 0, 0;
  --background-start-rgb: 255, 255, 255;
  --background-end-rgb: 255, 255, 255;
}

button,
[role="button"] {
  cursor: auto;
}

body {
  color: rgb(var(--foreground-rgb));
  background: linear-gradient(to bottom,
      transparent,
      rgb(var(--background-end-rgb))) rgb(var(--background-start-rgb));
}

@layer base {
  html {
    scroll-behavior: smooth;
  }
}