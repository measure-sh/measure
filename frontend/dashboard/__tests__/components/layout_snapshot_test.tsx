import LayoutSnapshot, { LayoutElementType, LayoutSnapshotStripedBgImage } from '@/app/components/layout_snapshot'
import { beforeEach, describe, expect, it, jest } from '@jest/globals'
import '@testing-library/jest-dom'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

type LayoutElement = {
    id: string
    label: string
    type: LayoutElementType
    x: number
    y: number
    width: number
    height: number
    scrollable: boolean
    highlighted: boolean
    children: LayoutElement[]
}

const mockElement: LayoutElement = {
    id: 'root',
    label: 'Root Container',
    type: 'container',
    x: 0,
    y: 0,
    width: 800,
    height: 600,
    scrollable: false,
    highlighted: false,
    children: []
}

const mockElementWithChildren: LayoutElement = {
    id: 'root',
    label: 'Root Container',
    type: 'container',
    x: 0,
    y: 0,
    width: 800,
    height: 600,
    scrollable: false,
    highlighted: false,
    children: [
        {
            id: 'child1',
            label: 'Button Element',
            type: 'button',
            x: 100,
            y: 100,
            width: 200,
            height: 50,
            scrollable: false,
            highlighted: false,
            children: []
        },
        {
            id: 'child2',
            label: 'Text Element',
            type: 'text',
            x: 400,
            y: 200,
            width: 300,
            height: 30,
            scrollable: false,
            highlighted: true,
            children: []
        }
    ]
}

const mockNestedElement: LayoutElement = {
    id: 'root',
    label: 'Root Container',
    type: 'container',
    x: 0,
    y: 0,
    width: 800,
    height: 600,
    scrollable: false,
    highlighted: false,
    children: [
        {
            id: 'parent',
            label: 'Parent Container',
            type: 'container',
            x: 50,
            y: 50,
            width: 400,
            height: 300,
            scrollable: false,
            highlighted: false,
            children: [
                {
                    id: 'nested-child',
                    label: 'Nested Input',
                    type: 'input',
                    x: 10,
                    y: 10,
                    width: 100,
                    height: 40,
                    scrollable: false,
                    highlighted: false,
                    children: []
                }
            ]
        }
    ]
}

const mockHorizontalElement: LayoutElement = {
    id: 'root',
    label: 'Root Container',
    type: 'container',
    x: 0,
    y: 0,
    width: 1000, // Wider than tall
    height: 400,
    scrollable: false,
    highlighted: false,
    children: []
}

global.fetch = jest.fn() as unknown as jest.Mock<
    (input: RequestInfo | URL, init?: RequestInit) => Promise<Response>
>

describe('LayoutSnapshot', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    const mockFetch = (data: LayoutElement) => {
        (global.fetch as jest.Mock<(input: RequestInfo | URL, init?: RequestInit) => Promise<Response>>)
            .mockResolvedValueOnce({
                ok: true,
                json: async () => data
            } as Partial<Response> as Response)
    }

    it('shows nothing while fetching', async () => {
        mockFetch(mockElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        // Should show nothing
        await waitFor(() => {
            expect(container.firstChild).toBeNull()
        })
    })

    it('renders correctly after fetching layout', async () => {
        mockFetch(mockElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })
    })

    it('calls fetch with correct URL', async () => {
        mockFetch(mockElement)

        render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith("http://example.com/layout.json")
        })
    })

    it('handles fetch errors gracefully', async () => {
        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => { });
        (global.fetch as jest.Mock<(input: RequestInfo | URL, init?: RequestInit) => Promise<Response>>)
            .mockRejectedValueOnce(new Error('Network error'))

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(consoleErrorSpy).toHaveBeenCalled()
        })

        // Should show nothing on error
        expect(container.firstChild).toBeNull()

        consoleErrorSpy.mockRestore()
    })

    it('renders with correct outer dimensions', async () => {
        mockFetch(mockElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            const wrapper = container.querySelector('.relative.overflow-hidden') as HTMLElement
            expect(wrapper).toBeInTheDocument()
        })
    })

    it('renders children elements correctly', async () => {
        mockFetch(mockElementWithChildren)

        render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            const elements = document.querySelectorAll('.absolute.border')
            expect(elements).toHaveLength(3) // root + 2 children
        })
    })

    it('applies highlighted styling to highlighted elements', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            const elements = container.querySelectorAll('.absolute.border')
            const highlightedChild = elements[2] // root, child1, child2 (highlighted)

            expect(highlightedChild).toHaveClass('border-yellow-200')
            expect(highlightedChild).toHaveStyle({
                backgroundImage: LayoutSnapshotStripedBgImage
            })
        })
    })

    it('handles hover interactions correctly', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const wrapper = container.querySelector('.relative')!
        fireEvent.mouseMove(wrapper, { clientX: 150, clientY: 200 })

        const elements = container.querySelectorAll('.absolute.border')
        const firstChild = elements[1]

        fireEvent.mouseEnter(firstChild)

        expect(screen.getByText('Button Element')).toBeInTheDocument()
    })

    it('shows and hides tooltip on mouse enter and leave', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const wrapper = container.querySelector('.relative')!
        fireEvent.mouseMove(wrapper, { clientX: 150, clientY: 200 })

        const elements = container.querySelectorAll('.absolute.border')
        const firstChild = elements[1]

        fireEvent.mouseEnter(firstChild)
        expect(screen.getByText('Button Element')).toBeInTheDocument()

        fireEvent.mouseLeave(firstChild)
        expect(screen.queryByText('Button Element')).not.toBeInTheDocument()
    })

    it('positions tooltip relative to mouse position', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const wrapper = container.querySelector('.relative')!
        const elements = container.querySelectorAll('.absolute.border')
        const firstChild = elements[1]

        fireEvent.mouseMove(wrapper, { clientX: 150, clientY: 200 })
        fireEvent.mouseEnter(firstChild)

        const tooltip = screen.getByText('Button Element')
        expect(tooltip).toHaveStyle({
            top: '220px', // clientY + 20
            left: '160px' // clientX + 10
        })
    })

    it('handles nested elements correctly', async () => {
        mockFetch(mockNestedElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            const elements = container.querySelectorAll('.absolute.border')
            expect(elements).toHaveLength(3) // root + parent + nested child
        })
    })

    it('finds and displays correct label for nested elements', async () => {
        mockFetch(mockNestedElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            const elements = container.querySelectorAll('.absolute.border')
            expect(elements).toHaveLength(3)

            const nestedChild = elements[2]
            expect(nestedChild).toHaveClass('border-gray-400')
        })
    })

    it('applies correct scaling for vertical orientation', async () => {
        // Mock element with vertical aspect ratio (height > width)
        const verticalElement: LayoutElement = {
            ...mockElement,
            width: 400,
            height: 800
        }

        mockFetch(verticalElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={211}
                height={366}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })
    })

    it('applies correct scaling for horizontal orientation', async () => {
        mockFetch(mockHorizontalElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={366}
                height={211}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })
    })

    it('handles empty children array', async () => {
        const elementWithEmptyChildren: LayoutElement = {
            ...mockElement,
            children: []
        }

        mockFetch(elementWithEmptyChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })
    })

    it('applies hover styling correctly', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const elements = container.querySelectorAll('.absolute.border')
        const firstChild = elements[1]

        expect(firstChild).toHaveClass('border-gray-400')

        fireEvent.mouseEnter(firstChild)
        expect(firstChild).toHaveClass('border-yellow-200')
    })

    it('maintains highlighted styling even when hovered', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const elements = container.querySelectorAll('.absolute.border')
        const highlightedChild = elements[2]

        expect(highlightedChild).toHaveClass('border-yellow-200')

        fireEvent.mouseEnter(highlightedChild)
        expect(highlightedChild).toHaveClass('border-yellow-200')
    })

    it('handles mouse movement without crashing', async () => {
        mockFetch(mockElement)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.relative')).toBeInTheDocument()
        })

        const wrapper = container.querySelector('.relative')!

        expect(() => {
            fireEvent.mouseMove(wrapper, { clientX: 100, clientY: 150 })
            fireEvent.mouseMove(wrapper, { clientX: 200, clientY: 250 })
        }).not.toThrow()
    })

    it('tooltip functionality works with proper sequence', async () => {
        mockFetch(mockElementWithChildren)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={800}
                height={600}
            />
        )

        await waitFor(() => {
            expect(container.querySelector('.absolute.border')).toBeInTheDocument()
        })

        const wrapper = container.querySelector('.relative')!
        const elements = container.querySelectorAll('.absolute.border')
        const firstChild = elements[1]

        fireEvent.mouseMove(wrapper, { clientX: 150, clientY: 200 })
        fireEvent.mouseEnter(firstChild)

        expect(screen.getByText('Button Element')).toBeInTheDocument()

        fireEvent.mouseMove(wrapper, { clientX: 200, clientY: 250 })

        const tooltip = screen.getByText('Button Element')
        expect(tooltip).toHaveStyle({
            top: '270px',
            left: '210px'
        })

        fireEvent.mouseLeave(firstChild)
        expect(screen.queryByText('Button Element')).not.toBeInTheDocument()
    })

    it('refetches layout when layoutUrl changes', async () => {
        mockFetch(mockElement)

        const { rerender } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout1.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith("http://example.com/layout1.json")
        })

        mockFetch(mockElementWithChildren)

        rerender(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout2.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith("http://example.com/layout2.json")
        })
    })

    it('handles non-ok response from fetch', async () => {
        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => { });
        (global.fetch as jest.Mock<(input: RequestInfo | URL, init?: RequestInit) => Promise<Response>>)
            .mockResolvedValueOnce({
                ok: false
            } as unknown as Partial<Response> as Response)

        const { container } = render(
            <LayoutSnapshot
                layoutUrl="http://example.com/layout.json"
                width={400}
                height={300}
            />
        )

        await waitFor(() => {
            expect(consoleErrorSpy).toHaveBeenCalled()
        })

        // Should show nothing on non-ok response
        expect(container.firstChild).toBeNull()

        consoleErrorSpy.mockRestore()
    })
})