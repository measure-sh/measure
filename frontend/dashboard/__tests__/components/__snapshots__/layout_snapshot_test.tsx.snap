// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutSnapshot handles empty children array 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative overflow-hidden"
        style="width: 400px; height: 300px;"
      >
        <div
          class="absolute border box-border cursor-pointer border-gray-400"
          style="left: 0px; top: 0px; width: 400px; height: 300px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative overflow-hidden"
      style="width: 400px; height: 300px;"
    >
      <div
        class="absolute border box-border cursor-pointer border-gray-400"
        style="left: 0px; top: 0px; width: 400px; height: 300px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LayoutSnapshot renders correctly with basic layout 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative overflow-hidden"
        style="width: 400px; height: 300px;"
      >
        <div
          class="absolute border box-border cursor-pointer border-gray-400"
          style="left: 0px; top: 0px; width: 400px; height: 300px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative overflow-hidden"
      style="width: 400px; height: 300px;"
    >
      <div
        class="absolute border box-border cursor-pointer border-gray-400"
        style="left: 0px; top: 0px; width: 400px; height: 300px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
